Use CheckedPtr for SharedTimer

rniwa opened the pull request:
#### c5c1a86c0f823abe547291692b3f5bd772799937
<pre>
Use CheckedPtr for SharedTimer
<a href="https://bugs.webkit.org/show_bug.cgi?id=292105">https://bugs.webkit.org/show_bug.cgi?id=292105</a>

Reviewed by Chris Dumez.

Use CheckedPtr&lt;SharedTimer&gt; in ThreadTimers instead of a raw pointer.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/platform/MainThreadSharedTimer.h:
* Source/WebCore/platform/SharedTimer.h:
* Source/WebCore/platform/ThreadTimers.cpp:
(WebCore::ThreadTimers::checkedSharedTimer):
(WebCore::ThreadTimers::setSharedTimer):
(WebCore::ThreadTimers::updateSharedTimer):
(WebCore::ThreadTimers::fireTimersInNestedEventLoop):
* Source/WebCore/platform/ThreadTimers.h:
* Source/WebCore/workers/WorkerRunLoop.cpp:
(WebCore::WorkerDedicatedRunLoop::WorkerDedicatedRunLoop):
(WebCore::RunLoopSetup::RunLoopSetup):
(WebCore::RunLoopSetup::~RunLoopSetup):

Canonical link: <a href="https://commits.webkit.org/294233@main">https://commits.webkit.org/294233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93f7b9483bf036d2bba62e70297d92a4e05fc11e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101233 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106381 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51860 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103274 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77100 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34134 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16350 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91439 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16166 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9449 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51208 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86051 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108737 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28361 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86075 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85625 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30354 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8065 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22460 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16462 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28291 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33562 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28103 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31423 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29661 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/ThreadTimers.cpp:
```diff
@@ -43,6 +43,11 @@ namespace WebCore {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(ThreadTimers);
 
+inline CheckedPtr<SharedTimer> ThreadTimers::protectedSharedTimer()
```
We usually call this `checkedSharedTimer()` I believe when returning a CheckedPtr.

rniwa replied:
Good point, will fix.

cdumez commented on Source/WebCore/workers/WorkerRunLoop.cpp:
```diff
@@ -133,8 +133,10 @@ class RunLoopSetup {
     {
-        if (!m_runLoop.m_nestedCount)
+        if (!m_runLoop.m_nestedCount) {
+            m_runLoop.m_sharedTimer = makeUnique<WorkerSharedTimer>();
```
Why did you move the construction here instead of the constructor?

rniwa replied:
Oh because the constructor runs in the main thread so thread identity check will fail otherwise.

cdumez commented on Source/WebCore/workers/WorkerRunLoop.cpp:
```diff
@@ -143,8 +145,10 @@ class RunLoopSetup {
-        if (!m_runLoop.m_nestedCount)
+        if (!m_runLoop.m_nestedCount) {
             threadGlobalData().threadTimers().setSharedTimer(nullptr);
+            m_runLoop.m_sharedTimer = nullptr;
```
And why are we newly nulling out here?

rniwa committed:
Use CheckedPtr for SharedTimer
https://bugs.webkit.org/show_bug.cgi?id=292105

Reviewed by Chris Dumez.

Use CheckedPtr<SharedTimer> in ThreadTimers instead of a raw pointer.

* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
* Source/WebCore/platform/MainThreadSharedTimer.h:
* Source/WebCore/platform/SharedTimer.h:
* Source/WebCore/platform/ThreadTimers.cpp:
(WebCore::ThreadTimers::checkedSharedTimer):
(WebCore::ThreadTimers::setSharedTimer):
(WebCore::ThreadTimers::updateSharedTimer):
(WebCore::ThreadTimers::fireTimersInNestedEventLoop):
* Source/WebCore/platform/ThreadTimers.h:
* Source/WebCore/workers/WorkerRunLoop.cpp:
(WebCore::WorkerDedicatedRunLoop::WorkerDedicatedRunLoop):
(WebCore::RunLoopSetup::RunLoopSetup):
(WebCore::RunLoopSetup::~RunLoopSetup):

Canonical link: https://commits.webkit.org/294233@main
